### ISSUE TITLE
[期权价差概率分析] | Spread Probability Analysis

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4519]
**标签**: quantitative-strategy
### 摘要
基于期权定价模型与概率统计，分析垂直价差组合的到期盈利概率与风险敞口。
### 算法描述
**核心原理**

垂直价差（Vertical Spread）由同一到期日、不同行权价的两个期权组成，核心在于量化价格变动触及价差边界的概率。

**Black-Scholes 定价模型**

期权价格由以下公式决定：
$$C = S_0 N(d_1) - K e^{-rT} N(d_2)$$

其中：
$$d_1 = \frac{\ln(S_0/K) + (r + \sigma^2/2)T}{\sigma\sqrt{T}}$$
$$d_2 = d_1 - \sigma\sqrt{T}$$

**概率分析框架**

1. **到期标的价格分布**：假设标的服从几何布朗运动，到期价格 $S_T$ 服从对数正态分布
$$S_T = S_0 \exp\left((r - \frac{\sigma^2}{2})T + \sigma\sqrt{T}Z\right), \quad Z \sim N(0,1)$$

2. **价差边界概率计算**：
   - 牛熊价差盈利概率：$P(S_T > K_{upper})$ 或 $P(S_T < K_{lower})$
   - 概率通过累积正态分布计算：$P = N\left(\frac{\ln(S_0/K) + (r \pm \sigma^2/2)T}{\sigma\sqrt{T}}\right)$

3. **期望收益计算**：
$$EV = P_{profit} \times Profit_{max} - P_{loss} \times Loss_{max}$$

4. **希腊字母风险度量**：
   - Delta：标的变动敏感度
   - Gamma：Delta变化率
   - Theta：时间衰减
   - Vega：波动率敏感度

**流程图**
```
输入参数 → BS定价 → 概率分布计算 → 期望收益 → 风险指标 → 输出推荐
```

### Python代码示例
```python
import numpy as np
from scipy.stats import norm
from scipy.optimize import brentq
from dataclasses import dataclass
from typing import Tuple, Optional
import matplotlib.pyplot as plt

@dataclass
class OptionParams:
    """期权参数"""
    S: float      # 标的当前价格
    K: float      # 行权价
    T: float      # 到期时间（年）
    r: float      # 无风险利率
    sigma: float  # 波动率
    q: float = 0.0  # 股息率

class BlackScholes:
    """Black-Scholes期权定价模型"""
    
    @staticmethod
    def d1_d2(S, K, T, r, sigma, q=0):
        """计算d1和d2"""
        d1 = (np.log(S / K) + (r - q + 0.5 * sigma**2) * T) / (sigma * np.sqrt(T))
        d2 = d1 - sigma * np.sqrt(T)
        return d1, d2
    
    @staticmethod
    def call_price(S, K, T, r, sigma, q=0):
        """看涨期权价格"""
        d1, d2 = BlackScholes.d1_d2(S, K, T, r, sigma, q)
        return S * np.exp(-q * T) * norm.cdf(d1) - K * np.exp(-r * T) * norm.cdf(d2)
    
    @staticmethod
    def put_price(S, K, T, r, sigma, q=0):
        """看跌期权价格"""
        d1, d2 = BlackScholes.d1_d2(S, K, T, r, sigma, q)
        return K * np.exp(-r * T) * norm.cdf(-d2) - S * np.exp(-q * T) * norm.cdf(-d1)


class SpreadProbabilityAnalyzer:
    """价差概率分析器"""
    
    def __init__(self, params: OptionParams):
        self.params = params
        self.bs = BlackScholes()
    
    def calc_probability_profit(self, K_long: float, K_short: float, 
                                 is_bull: bool = True) -> dict:
        """
        计算价差盈利概率
        
        Args:
            K_long: 买入期权的行权价（对于牛市价差是较低的价）
            K_short: 卖出期权的行权价
            is_bull: True=牛市价差，False=熊市价差
        """
        S, T, r, sigma = self.params.S, self.params.T, self.params.r, self.params.sigma
        
        if is_bull:
            # 牛市看涨价差：买入低行权价K_long，卖出高行权价K_short
            # 盈利区间：S_T > K_short
            d1 = (np.log(S / K_short) + (r + 0.5 * sigma**2) * T) / (sigma * np.sqrt(T))
            prob_profit = norm.cdf(d1)
            
            # 最大盈利
            call_long = self.bs.call_price(S, K_long, T, r, sigma)
            call_short = self.bs.call_price(S, K_short, T, r, sigma)
            max_profit = (K_short - K_long) - (call_long - call_short) * np.exp(r * T)
            max_loss = call_long - call_short
        else:
            # 熊市看涨价差：卖出低行权价K_short，买入高行权价K_long
            # 盈利区间：S_T < K_short
            d1 = (np.log(S / K_short) + (r + 0.5 * sigma**2) * T) / (sigma * np.sqrt(T))
            prob_profit = norm.cdf(-d1)
            
            call_long = self.bs.call_price(S, K_long, T, r, sigma)
            call_short = self.bs.call_price(S, K_short, T, r, sigma)
            max_profit = call_short - call_long
            max_loss = (K_long - K_short) - (call_short - call_long) * np.exp(r * T)
        
        ev = prob_profit * max_profit - (1 - prob_profit) * max_loss
        
        return {
            'prob_profit': prob_profit,
            'max_profit': max_profit,
            'max_loss': max_loss,
            'expected_value': ev,
            'reward_risk_ratio': max_profit / max_loss if max_loss > 0 else np.inf
        }
    
    def analyze_vertical_spread(self, spread_type: str, 
                                 strikes: Tuple[float, float]) -> dict:
        """
        综合分析垂直价差
        
        Args:
            spread_type: 'bull_call', 'bear_call', 'bull_put', 'bear_put'
            strikes: (lower_strike, upper_strike)
        """
        K1, K2 = min(strikes), max(strikes)
        results = {}
        
        if spread_type in ['bull_call', 'bear_call']:
            # 看涨价差
            if spread_type == 'bull_call':
                result = self.calc_probability_profit(K1, K2, is_bull=True)
                result['strategy'] = 'Bull Call Spread'
            else:
                result = self.calc_probability_profit(K2, K1, is_bull=False)
                result['strategy'] = 'Bear Call Spread'
            results = result
        else:
            # 看跌价差
            call_long = self.bs.call_price(self.params.S, K1, self.params.T, 
                                          self.params.r, self.params.sigma)
            call_short = self.bs.call_price(self.params.S, K2, self.params.T, 
                                           self.params.r, self.params.sigma)
            
            if spread_type == 'bull_put':
                # 牛市看跌价差：卖出高行权价K2，买入低行权价K1
                net_credit = call_short - call_long
                # 盈利区间：S_T > K2
                d1 = (np.log(self.params.S / K2) + 
                      (self.params.r + 0.5 * self.params.sigma**2) * self.params.T) / \
                     (self.params.sigma * np.sqrt(self.params.T))
                prob_profit = norm.cdf(d1)
                max_profit = net_credit
                max_loss = (K2 - K1) - net_credit
                results = {
                    'strategy': 'Bull Put Spread',
                    'prob_profit': prob_profit,
                    'max_profit': max_profit,
                    'max_loss': max_loss,
                    'expected_value': prob_profit * max_profit - (1 - prob_profit) * max_loss,
                    'reward_risk_ratio': max_profit / max_loss if max_loss > 0 else np.inf
                }
            else:
                # 熊市看跌价差：买入高行权价K2，卖出低行权价K1
                net_debit = call_long - call_short
                # 盈利区间：S_T < K1
                d1 = (np.log(self.params.S / K1) + 
                      (self.params.r + 0.5 * self.params.sigma**2) * self.params.T) / \
                     (self.params.sigma * np.sqrt(self.params.T))
                prob_profit = norm.cdf(-d1)
                max_loss = net_debit
                max_profit = (K2 - K1) - net_debit
                results = {
                    'strategy': 'Bear Put Spread',
                    'prob_profit': prob_profit,
                    'max_profit': max_profit,
                    'max_loss': max_loss,
                    'expected_value': prob_profit * max_profit - (1 - prob_profit) * max_loss,
                    'reward_risk_ratio': max_profit / max_loss if max_loss > 0 else np.inf
                }
        
        return results
    
    def find_optimal_spread(self, target_prob: float = 0.65) -> dict:
        """寻找满足目标概率的最优价差"""
        S, T, r, sigma = self.params.S, self.params.T, self.params.r, self.params.sigma
        
        # 波动率微笑/skew调整
        K_otm = S * np.exp(sigma * np.sqrt(T))
        strikes = np.linspace(S * 0.8, S * 1.2, 50)
        
        best_spread = None
        best_ev = -np.inf
        
        for i, K1 in enumerate(strikes):
            for K2 in strikes[i+1:]:
                result = self.calc_probability_profit(K1, K2, is_bull=True)
                if abs(result['prob_profit'] - target_prob) < 0.02:
                    if result['expected_value'] > best_ev:
                        best_ev = result['expected_value']
                        best_spread = {
                            'K_long': K1, 'K_short': K2,
                            **result
                        }
        
        return best_spread


if __name__ == '__main__':
    # 示例：分析苹果股票牛市看涨价差
    params = OptionParams(
        S=175.0,      # 苹果当前价格
        K=175.0,      # ATM
        T=30/365,     # 30天到期
        r=0.05,       # 5%无风险利率
        sigma=0.25    # 25%波动率
    )
    
    analyzer = SpreadProbabilityAnalyzer(params)
    
    # 牛市看涨价差分析
    result = analyzer.analyze_vertical_spread('bull_call', (170.0, 180.0))
    print(f"策略: {result['strategy']}")
    print(f"盈利概率: {result['prob_profit']:.2%}")
    print(f"最大盈利: ${result['max_profit']:.2f}")
    print(f"最大亏损: ${result['max_loss']:.2f}")
    print(f"期望收益: ${result['expected_value']:.2f}")
    print(f"盈亏比: {result['reward_risk_ratio']:.2f}")
```

### 数据来源
- Reddit: r/quant, r/algorithmictrading
- 小红书: 量化投资
- 知乎: 量化交易
